File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        z = []
        for i in range(len(nums)-2):
            k=i+1
            j=len(nums)-1
            while k<j:
                if nums[i]+nums[j]+nums[k]>0:
                    j-=1
                elif nums[i]+nums[j]+nums[k]<0:
                    k+=1
                elif nums[i]+nums[j]+nums[k]==0:
                    z.append((nums[i],nums[j],nums[k]))
                    k+=1
        return list(set(z))

File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        int n = nums.size();

        sort(nums.begin(), nums.end());

        for (int i = 0; i < n - 2; i++) {

            if (i > 0 and nums[i] == nums[i-1]) continue;
            for (int j = i + 1, k = n - 1; j < k; ) {

                int val = nums[i] + nums[j] + nums[k];
                if (val == 0) {
                    ans.push_back({nums[i], nums[j], nums[k]});
                    j++;
                    while (nums[j] == nums[j - 1] and j < k) j++;
                }
                else if (val > 0) k--;
                else j++;
            }
        }

        return ans;
    }
};